internal/config/cgroupmanager/supported.go

```diff
@@ -0,0 +1,123 @@
+// +build linux
+
+package cgroupmanager
+
+import (
+	"io/ioutil"
+	"os"
+	"strings"
+	"sync"
+
+	libpodcgroups "github.com/containers/libpod/pkg/cgroups"
+	libctrcgroups "github.com/opencontainers/runc/libcontainer/cgroups"
+	"github.com/pkg/errors"
+)
+
+var (
+	_cgroupHasHugetlbOnce sync.Once
+	_cgroupHasHugetlb     bool
+	_cgroupHasHugetlbErr  error
+
+	_cgroupHasPidOnce sync.Once
+	_cgroupHasPid     bool
+	_cgroupHasPidErr  error
+
+	_cgroupHasMemorySwapOnce sync.Once
+	_cgroupHasMemorySwap     bool
+	_cgroupHasMemorySwapErr  error
+
+	_cgroupIsV2Once sync.Once
+	_cgroupIsV2     bool
+	_cgroupIsV2Err  error
+)
+
+// initializeCgroups initializes all of the singleton variables
+// that store the node's cgroup configuration. Currently, we check
+// hugetlb, cgroup v1 or v2, pid and memory swap support
+// We check the error at server initialization, and if we error, shutdown
+// cri-o early, instead of when we're already trying to run containers.
+func initializeCgroups() error {
+	toInit := []struct {
+		initializer func() bool
+		err         *error
+	}{
+		{
+			initializer: CgroupHasHugetlb,
+			err:         &_cgroupHasHugetlbErr,
+		},
+		{
+			initializer: CgroupHasPid,
+			err:         &_cgroupHasPidErr,
+		},
+		{
+			initializer: CgroupIsV2,
+			err:         &_cgroupIsV2Err,
+		},
+		{
+			initializer: CgroupHasMemorySwap,
+			err:         &_cgroupHasMemorySwapErr,
+		},
+	}
+	for _, i := range toInit {
+		i.initializer()
+		if *i.err != nil {
+			return errors.Wrapf(*i.err, "failed to initialize cgroup information")
+		}
+	}
+	return nil
+}
+
+// CgroupHasHugetlb returns whether the hugetlb controller is present
+func CgroupHasHugetlb() bool {
+	_cgroupHasHugetlbOnce.Do(func() {
+		if CgroupIsV2() {
+			if _, err := ioutil.ReadDir("/sys/fs/cgroup/hugetlb"); err != nil {
+				_cgroupHasHugetlbErr = errors.Wrap(err, "readdir /sys/fs/cgroup/hugetlb")
+				return
+			}
+			_cgroupHasHugetlb = true
+			return
+		}
+
+		controllers, err := ioutil.ReadFile("/sys/fs/cgroup/cgroup.controllers")
+		if err != nil {
+			_cgroupHasHugetlbErr = errors.Wrap(err, "read /sys/fs/cgroup/cgroup.controllers")
+			return
+		}
+		_cgroupHasHugetlb = strings.Contains(string(controllers), "hugetlb")
```

Comment:
this can result in false positive if some future kernel will introduce say `hugetlb2`.

Reply:
@kolyshkin do you have a patch that you think is better? I'm not super inclined update the check to something more complicated (using regexp or something) unless the future of hugetlb2 feels realistic. Even then, considering how slowly the kernel moves, I would imagine we would have plenty of time to cross the bridge once we get there. This line is clear and comprehensive for now, which I prefer, unless you have something better in mind.

Reply:
Realistically, you're right. And yet most of the implementations I've dealt with would do something like
```go
for _, ctr := range strings.Split(string(controllers)) {
  if ctr == "hugetlb" {
    __cgroupHasHugetlb = true
    break
  }
}
```
A simple non-regex solution would be
```go
_cgroupHasHugetlb = strings.Contains(" "+string(controllers)+" ", " hugetlb ")
```

Reply:
okay I've adapted the first version in the new generic controller population code (as suggested below)